### DISCUSSION TITLE
Custom scopes?

### DISCUSSION THREAD

**Discussion by susliko** [VirtusLab/scala-cli#3124] [Ideas]
As far as I understand, scala-cli directives can define two scopes: `main` and `test`.
I can think of at least of the following use case that might leverage an ability to define custom scopes.

## Publish a subset of sources in a workspace
Currently `publish` command bundles all source with non-test scopes into a jar.
But it might be useful to narrow a set o sources for publish.
For example, in the following structure for some micro-library:

```
main/ #  Library sources. Should be published
test/ # Library tests. Should not be published
examples/ # Code samples that just show usages of a library
project.scala
publishing-conf.scala
```
In this structure it would be great to be able to execute `scala-cli publish` and have only sources from `main` bundled into a jar.
Currently the workaround is `scala-cli publish main/ *.scala`

**Comment by Gedochao** [comment]
So this is likely a valid feature request and has been discussed in the past. 
Thing is, implementation of custom scopes defined by the user would likely be quite complex.
Feel free to convert this to an issue (or open a new one) to the backlog.
However, it is unlikely this would be worked on by the maintenance team in the predictable future. 
The door is open for external contributions.

**Reply by susliko** [reply]
@Gedochao Thanks!
Can you think of other possible use cases for custom scopes and its semantics with respect to commands other than `publish`?
I might try implementing it whenever I'll have some time.

**Reply by Gedochao** [reply]
Hey, sorry for the late answer.
Potential other use cases which come to mind are benchmarking and markdown builds.
